### PR TITLE
Optimize 074 dump all

### PR DESCRIPTION
**Pull request by litghost**
Variant of #1220 with some attempts at fixing failures.

### CODE REVIEW

**Comment by acomodi** [comment]
Artix failed on extra harness targets, got a fix in the original PR that is rebased on top of this one: https://github.com/SymbiFlow/prjxray/pull/1220.

Being `run.ok` already present in the fuzzers, they weren't run to produce the new yaml and json files for the harnesses.

**Comment by litghost** [comment]
New total runtime is ~9 hrs, but the 50T and 200T fuzzers run serially.  If we modify some of the fuzzers to run the A200T in parallel, I believe we can get the runtime to ~5 hrs, which is likely good enough.

**Comment by acomodi** [comment]
@litghost I have also noticed an additional issue.

The 200T has new kind of tiles that gets added, e.g. `artix7/tile_type_GTP_CHANNEL_0_MID_LEFT.json`.

I believe it is not an issue to have them all together with the ones related to the 50T only, but this may generate confusion.

Would it be worth to add the tile_type and site_type files also in the part-dependant directories?

**Comment by litghost** [comment]
> @litghost I have also noticed an additional issue.
> 
> The 200T has new kind of tiles that gets added, e.g. `artix7/tile_type_GTP_CHANNEL_0_MID_LEFT.json`.
> 
> I believe it is not an issue to have them all together with the ones related to the 50T only, but this may generate confusion.
> 
> Would it be worth to add the tile_type and site_type files also in the part-dependant directories?

This doesn't bother me so much.  I believe there is some cleanup that can be done in a bit where we merge the site, tile and segbits across parts.  Tilegrid and tileconn are part specific, but I believe we have quiet a bit of evidence that tiles, sites, and segbits are generic across the entire 7-series line.  We kept them seperate in case of exceptions, but to my knowledge, we haven't found any exceptions yet?

**Comment by acomodi** [comment]
Well, what made me think this may was an issue is that there are differences in some of the common tile_types:

```
 artix7/tile_type_BRKH_INT.json                     |     56 +-
 artix7/tile_type_CFG_CENTER_MID.json               |    110 +-
 artix7/tile_type_CFG_CENTER_TOP.json               |    120 +-
 artix7/tile_type_CLBLL_L.json                      |    188 +-
 artix7/tile_type_CLBLM_L.json                      |    692 +-
 artix7/tile_type_CLBLM_R.json                      |      2 +-
 artix7/tile_type_CLK_BUFG_BOT_R.json               |    330 +-
 artix7/tile_type_CLK_BUFG_TOP_R.json               |    330 +-
 artix7/tile_type_CLK_HROW_BOT_R.json               |    382 +-
 artix7/tile_type_CLK_HROW_TOP_R.json               |    382 +-
 artix7/tile_type_CMT_FIFO_L.json                   |   1238 +-
```

I believe this is due to the newly added sorting though, so I believe that your statement is correct that tiles and sites are generic across 7-series

**Comment by litghost** [comment]
> newly added sorting though

I believe so to.  I have some changes to the sorting that might help, working on it in #1237

**Comment by litghost** [comment]
> New total runtime is ~9 hrs, but the 50T and 200T fuzzers run serially. If we modify some of the fuzzers to run the A200T in parallel, I believe we can get the runtime to ~5 hrs, which is likely good enough.
> 
> The following fuzzers would need modification:
> 
> * 001
> * 005
> * 072
> * 074
> 
> I'll dig a little to see how much effort is required to achieve this.

I believe I have a surgical change to enable this, testing now.

**Comment by acomodi** [comment]
```
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:   File "add_tdb.py", line 140, in <module>
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:     main()
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:   File "add_tdb.py", line 136, in main
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:     run(args.fn_in, args.fn_out, verbose=args.verbose)
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:   File "add_tdb.py", line 112, in run
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:     for (tile, frame, wordidx) in load_db(tdb_fn):
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:   File "add_tdb.py", line 40, in load_db
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:     check_frames(tagstr, addrlist)
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:   File "add_tdb.py", line 11, in check_frames
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:     frame = parse_addr(addrstr, get_base_frame=True)
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:   File "add_tdb.py", line 20, in parse_addr
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s:     frame = int(line[0], 16)
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s: ValueError: invalid literal for int() with base 16: '<const0>'
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s: make[3]: *** [build_xc7z020clg484-1/tilegrid_tdb.json] Error 1
2020-02-18T17:47:33.917847 - 005-tilegrid              - 9m06s: make[2]: *** [run] Error 2
```

It looks like there is an instability of the xc7z020. This is in particular related to clb_int in 005-tilegrid fuzzer. I will file an issue, as this kinds of instability should get fixed.

**Comment by litghost** [comment]
@acomodi This PR has had significant changes since last review, so it probably worth a re-review.

File: .github/kokoro/db-full.sh [hunk 1]
```diff
@@ -69,37 +60,9 @@ echo "----------------------------------------"
 		rm $tmp
 		exit $DATABASE_RET
 	fi
-
```
Comment (mithro) [inline .github/kokoro/db-full.sh]: Why are you removing the check which makes sure that there is nothing left to run after a complete fuzzing run?

File: fuzzers/000-init-db/Makefile [hunk 2]
```diff
@@ -82,15 +84,18 @@ status:
...
-pushdb: $(DB_FILES_PATH)
+${XRAY_FAMILY_DIR}/${XRAY_PART}:
+	mkdir -p ${XRAY_FAMILY_DIR}/${XRAY_PART}
+
+pushdb: $(DB_FILES_PATH) ${XRAY_FAMILY_DIR}/${XRAY_PART}
 	@true
 
 run:
-	$(MAKE) clean
+	rm -rf $(BUILD_DIR) run.${XRAY_PART}.ok
```
Comment (mithro) [inline fuzzers/000-init-db/Makefile]: Should there be a
```makefile

clean-${XRAY_PART}:
	rm -rf $(BUILD_DIR) run.${XRAY_PART}.ok

run:
	$(MAKE) clean-${XRAY_PART}
```

File: fuzzers/005-tilegrid/add_tdb.py [hunk 3]
```diff
@@ -76,34 +78,35 @@ def run(fn_in, fn_out, verbose=False):
...
-        ("bram_int/build/segbits_tilegrid.tdb", int_frames, int_words),
-        ("dsp_int/build/segbits_tilegrid.tdb", int_frames, int_words),
-        ("fifo_int/build/segbits_tilegrid.tdb", int_frames, int_words),
-        ("ps7_int/build/segbits_tilegrid.tdb", int_frames, int_words),
-        ("cfg_int/build/segbits_tilegrid.tdb", int_frames, int_words),
-        ("monitor_int/build/segbits_tilegrid.tdb", int_frames, int_words),
-        (
-            "orphan_int_column/build/segbits_tilegrid.tdb", int_frames,
-            int_words),
+        ("iob", 42, 4),
```
Comment (mithro) [inline fuzzers/005-tilegrid/add_tdb.py]: I know you didn't add these in this CL, but it would be nice to have a comment about why some are hard coded and why some are not?

File: fuzzers/Makefile [hunk 4]
```diff
@@ -19,161 +19,169 @@ clean_locks:
...
 ifneq ($(QUICK),Y)
-$(eval $(call fuzzer,058-pip-hclk,005-tilegrid))
-$(eval $(call fuzzer,059-pip-byp-bounce,048-int-piplist))
-$(eval $(call fuzzer,060-bram-cascades,005-tilegrid))
-$(eval $(call fuzzer,071-ppips,057-pip-bi 058-pip-hclk 060-bram-cascades))
+$(eval $(call fuzzer,058-pip-hclk,005-tilegrid,N))
+$(eval $(call fuzzer,059-pip-byp-bounce,048-int-piplist,N))
+$(eval $(call fuzzer,060-bram-cascades,005-tilegrid,N))
+$(eval $(call fuzzer,071-ppips,057-pip-bi 058-pip-hclk 060-bram-cascades,N))
 ifneq ($(BITONLY),Y)
```
Comment (mithro) [inline fuzzers/Makefile]: Should the following still be `Y`?

File: .github/kokoro/db-full.sh (hunk 1 above)
Comment (litghost) [reply .github/kokoro/db-full.sh]: Because the check is wrong when there are multiple sub-directories.  The Makefile refactoring resulted in multiple make sub-direction recursions, which this code cannot handle.  There is not a straight forward replacement that isn't fragile, like this previous code was.

File: fuzzers/Makefile (hunk 4 above)
Comment (litghost) [reply fuzzers/Makefile]: Which line are you refering too?  I don't follow what you are referring too.

File: fuzzers/005-tilegrid/add_tdb.py (hunk 3 above)
Comment (litghost) [reply fuzzers/005-tilegrid/add_tdb.py]: All of the int frame TDB's are really the same constant, hence the variable `int_frames`/`int_words`.  All the others are ones offs and vary.

File: fuzzers/000-init-db/Makefile (hunk 2 above)
Comment (litghost) [reply fuzzers/000-init-db/Makefile]: I'd argue no.  If you are running interactively, `make clean` does the right thing. If you are doing multiple parts, you won't typically be looking at one fuzzer.

File: .github/kokoro/db-full.sh (hunk 1 above)
Comment (mithro) [reply .github/kokoro/db-full.sh]: Shouldn't `make --dry-run db-${XRAY_SETTINGS}-all` check that the `make db-${XRAY_SETTINGS}-all` target doesn't have anything left to do? Or is `db-${XRAY_SETTINGS}-all` a phony target so `make` always returns something when using `--dry-run`?

Comment (litghost) [reply .github/kokoro/db-full.sh]: > Or is `db-${XRAY_SETTINGS}-all` a phony target so `make` always returns something when using `--dry-run`?

This, sort of.  Each leaf phony target reports "nothing to do", but there are bunch of "entering directory", "leaving directory" and the print of the recursive make target invocation.  The end result is it is fairly hard to parse the output.

File: fuzzers/Makefile (hunk 4 above)
Comment (mithro) [reply fuzzers/Makefile]: The last value in the `$(eval)` is the verbose setting, right? Should it still be verbose for these fuzzers?
```
$(eval $(call fuzzer,072-ordered_wires,,Y))
$(eval $(call fuzzer,073-get_counts,))	$(eval $(call fuzzer,073-get_counts,,Y))
$(eval $(call fuzzer,074-dump_all,005-tilegrid 072-ordered_wires))	$(eval $(call fuzzer,074-dump_all,005-tilegrid 072-ordered_wires,Y))
$(eval $(call fuzzer,075-pins,))	$(eval $(call fuzzer,075-pins,,Y))
```

File: fuzzers/Makefile [hunk 5]
```diff
@@ -19,161 +19,169 @@ clean_locks:
...
 clean_logs::
-	rm -rf $(1)/logs
+	rm -rf $(1)/logs_*
 
 # Describe how to create the fuzzer's run.ok file.
 # This command must start with a + to tell make to pass the jobserver
 # parameters downwards.
 ifeq ($(VERBOSE),Y)
 
 # When verbose we just call make directory
```
Comment (mithro) [inline fuzzers/Makefile]: Add new line here.

File: fuzzers/Makefile (hunk 4 above)
Comment (litghost) [reply fuzzers/Makefile]: Fixed.

File: fuzzers/005-tilegrid/cfg_int/Makefile [hunk 6]
```diff
@@ -1,3 +1,3 @@
-N ?= 4
+N ?= 10
```
[the remark this answers is not in the document]
Comment (litghost) [reply fuzzers/005-tilegrid/cfg_int/Makefile]: This was causing some fuzzer instability.  It might be the cause of https://github.com/SymbiFlow/prjxray/pull/1227?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDcxMDIxNDIxODo1MzcwNzQ%3D&notifications_query=is%3Aunread#issuecomment-587601068

File: fuzzers/005-tilegrid/add_tdb.py [hunk 7]
```diff
@@ -37,6 +38,7 @@ def load_db(fn):
         parts = l.split(' ')
         tagstr = parts[0]
         addrlist = parts[1:]
+        assert not any(s == '<const0>' for s in addrlist), (fn, l)
```
[the remark this answers is not in the document]
Comment (litghost) [reply fuzzers/005-tilegrid/add_tdb.py]: That's good future work

**Comment by litghost** [comment]
I think I've identified a bug due to the excluded tiles in 074 locally, investigating now.
